### DISCUSSION TITLE
Perf: better to require autoloader inside or outside worker loop?

### DISCUSSION THREAD

**Discussion by summersab** [roadrunner-server/roadrunner#1531] [General]
I'm wanting to work on adding support for RoadRunner to a major project (Nextcloud). This project has several thousand classes in its autoloader. This made me wonder - in terms of performance, is it better to have the autoloader called outside the worker loop or does it not really matter? It would be somewhat easier to call the autoloader inside the worker loop based on the current Nextcloud structure. However, if there is a performance benefit, that information would be helpful so I can justify proposing code changes.

Thanks!

**Comment by rauanmayemir** [comment]
Does it really matter if you use opcode cache?

**Reply by rustatian** [reply]
CC: @butschster @roxblnfk @wolfy-j

**Reply by summersab** [reply]
I don't know. I'm not exactly familiar with the finer details of the inner workings of PHP. I just know how to write code and set up servers using recommended settings. I suppose I considered this a "recommended setting," and since I found no mention of the best practice, I thought I'd ask.

(...or were you asking that question of those whom you CC'd? I thought you were replying to me with a rhetorical question).

**Reply by wolfy-j** [reply]
OPCache provides a little benefit under RoadRunner, same with loading code eagerly. It is the most effective way to load code once and then work with it from inside the loop (for example in Spiral Framework we specifically made bootloading phase heavy of computations to reduce the load on runtime phase). 

The only issue you have to be aware, keeping LARGE codebase in memory would obviously require more memory, it's cheap these days, but still. 

In your ideal setup you want to make sure that application has no code related IO during it's operation phase (i.e. serving requests), all configs are already in RAM, same as code. So your disk usage (which is the slowest memory storage) won't be used, but your RAM (which is the fastest) will be utilized as MAX.

Try to keep ONLY runtime code inside the loop by memorizing stuff ahead of the loop and it will speed up your application even more. I'm not aware which type of architecture or framework you use, but if it's possible go hard on making shared services and singletons since they only have to created once per bootloading (i.e. right after the app starts).

**Reply by rauanmayemir** [reply]
Indeed, it was a rhetorical question. 😄 

I actually meant to say preload, not regular opcode cache. https://stitcher.io/blog/preloading-in-php-74
I believe it actually targets use cases like yours.

**Reply by wolfy-j** [reply]
You can find some diagrams explaining the working model here - https://spiral.dev/docs/framework-lifecycle/3.6/en

**Reply by wolfy-j** [reply]
> Indeed, it was a rhetorical question. 😄
> 
> I actually meant to say preload, not regular opcode cache. https://stitcher.io/blog/preloading-in-php-74 I believe it actually targets use cases like yours.

Preloading also have little effect, it will only speed up the bootstrapping phase which you do once. Nothing can really beat having code in RAM.

**Reply by wolfy-j** [reply]
However, you do not need to FORCE loading code in your worker bootload phase. You can keep reading classes you need as you go, it called worker "warming up" and it will take some time for worker to reach it's optimal memory state and performance, but it also much easier to implement since it requires no changes in autoloading.

**Reply by summersab** [reply]
This is all excellent information - thank you! I may revisit this and ask more questions down the road if I have more questions. Adding RR support to Nextcloud is a side project for me at the moment.

Is there documentation regarding best practices for RR, how to structure projects and code to improve performance, and what settings to tune and tweak? I've seen a few posts regarding the number of workers and RAM, but I haven't seen much regarding best practices from an actual coding perspective. Things like:
- What should/shouldn't be put in the worker loop
- How best to efficiently load the application into memory to improve worker performance
- The operations that should be avoided because they have a negative impact on performance
- Tradeoffs and pitfalls, etc.

A nice bulleted checklist on a wiki page would be a great starting point.

**Reply by rustatian** [reply]
We have this advice for a production use: https://roadrunner.dev/docs/app-server-production/2.x/en
